src/url.rs

```diff
@@ -177,15 +177,25 @@ impl PyUrl {
         };
         let mut url = format!("{scheme}://{url_host}");
         if let Some(path) = path {
-            url.push('/');
-            url.push_str(path);
+            if !url.ends_with('/') {
+                url.push('/');
+            }
+            if path.starts_with('/') {
+                url.push_str(path.trim_start_matches('/'));
+            } else {
+                url.push_str(path);
+            }
         }
         if let Some(query) = query {
-            url.push('?');
+            if !query.starts_with('?') {
+                url.push('?');
+            }
```

Comment:
This seems like a change in functionality which is slightly different, so perhaps needs to be split out into a separate PR. Similar to my comment about `hostname`, there should probably be some validation here.